### PR TITLE
Convert user-uploaded animated GIFs to videos

### LINKED ISSUE #7232: Convert user uploaded animated GIFs to videos
## Feature Description

Since we now have all the pieces to do use FFmpeg in the editor and force cross-origin isolation, it would totally be possible to optimize a user-uploaded animated GIF by automatically turning it into an MP4. In the editor this would use the "hidden" GIF element we already use for Tenor GIFs (which are also actually videos).

Not only would it be possible, but it should also be quite trivial to implement:

During upload:

1. Upload GIF file
2. We check whether the GIF is animated (perhaps with something like https://gist.github.com/marckubischta/261ad8427a214022890b)
3. If GIF is animated, run through FFmpeg
4. Create GIF element with the video
5. Create poster image for the video

Existing images:

1. Just like for non-optimized videos, offer a button to optimize GIF, which will optimize it behind the scenes

## Alternatives Considered

This might not be a high priority since not many users might be uploading animated GIFs to WordPress / stories.

## Additional Context

Related: https://github.com/WordPress/gutenberg/issues/19730

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

Converts any user-uploaded animated GIF to a video before uploading.

Does not affect static GIFs.

Images upload via the WP media dialog are also converted. 

## Relevant Technical Choices

This feature is behind a feature flag. 

## To-do

* [x] As there are some differences between the shape of a GIF `resource` coming from WP and the ones coming from `Tenor`, this probably needs a data migration to unify the format.

## User-facing changes

https://user-images.githubusercontent.com/237508/122044800-d7126580-cdd4-11eb-92d5-0e0c52d15add.mov

### QA

- [ ] This is a non-user-facing change and requires no QA

**Prerequisites:**

Enable GIF optimization on the Experiments page.

**Files for testing**

[GIFs for testing.zip](https://github.com/google/web-stories-wp/files/6775705/GIFs.for.testing.zip)

**Instructions**

This PR can be tested by following these steps:

**Upload via media modal**

1. Click on `Upload` button to open WordPress media modal
2. Upload `nyancat.gif`
3. Insert GIF to page
4. Verify that video optimization starts and GIF is converted to video
5. Save story
6. Refresh window
7. Verify that GIF is displayed in media library with a poster image & plays on hover
5. Preview story
6. Verify that GIF plays as expected

**Upload via drag & drop**

1. Upload `nyancat.gif` via drag & drop to editor
4. Verify that video optimization starts and GIF is converted to video
5. Save story
6. Refresh window
7. Verify that GIF is displayed in media library with a poster image & plays on hover
5. Preview story
6. Verify that GIF plays as expected

**Still image**

1. Upload `still.gif` via drag & drop to editor
2. Verify that GIF is not converted to video, and that it stays a still image that can be used right away

**Tenor GIFs**

1. Go to the third-party media tab
2. Insert a GIF
3. Verify that GIF plays on the canvas as expected
4. Preview story
5. Verify that GIF plays as expected

### UAT

- [x] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #7232

### CODE REVIEW

**Comment by spacedmonkey** [comment]
You maybe able to do something clever like I have done in [here](https://github.com/google/web-stories-wp/pull/7766) and trigger a convertion from gif -> mp4 on insert from media picker.

File: assets/src/edit-story/app/media/utils/useProcessVideo.js [hunk 1]
```diff
@@ -105,7 +105,7 @@ function useProcessVideo({
         updateOldVideo(oldResource.id, resource.id);
         deleteMediaElement({ id: oldResource.id });
 
-        if (resource.type === 'video' && !resource.local) {
+        if (['video', 'gif'].includes(resource.type) && !resource.local) {
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/media/utils/useProcessVideo.js:108]: Don't we need to check if this is tenor?

File: assets/src/edit-story/app/media/utils/getResourceFromLocalFile.js [hunk 2]
```diff
@@ -60,9 +45,7 @@ const getImageResource = async (file) => {
   const fileName = getFileName(file);
   const mimeType = file.type;
 
-  const reader = await createFileReader(file);
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/media/utils/getResourceFromLocalFile.js]: Why was this removed?

File: assets/src/edit-story/app/media/local/useContextValueProvider.js [hunk 3]
```diff
@@ -167,9 +167,9 @@ export default function useContextValueProvider(reducerState, reducerActions) {
   });
 
   const generateMissingPosters = useCallback(
-    ({ mimeType, posterId, id, src, local }) => {
+    ({ mimeType, posterId, id, src, local, type }) => {
       if (
-        allowedVideoMimeTypes.includes(mimeType) &&
+        (allowedVideoMimeTypes.includes(mimeType) || type === 'gif') &&
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/media/local/useContextValueProvider.js:172]: Don't we need to check if gif and video version has been generated?

**Comment by spacedmonkey** [comment]
@googlebot I consent.

**Comment by spacedmonkey** [comment]
I have some time getting up to speed on this PR, I have made the following tweaks. 

- Tweaked how poster generation works. 
- Restored preview on drag into canvas. 
- Added feature flag. 
- Add logic to convert gifs inserted via media library. 

What I have an issue with now is UX. If I user, drags and gif file into canvas, do I expect that file to become a video? The file is converted gif to an mp4 and then uploaded. The gif file is never uploaded and WordPress and the editor treat it as a video file. Is that confusing? i would not be able to filter by image to find the image. It is also worth noting that there are no ways to filter by gif in the editor. Unlike third party libraries there is only an option for image / video. 

I am wondering, if this PR should have a different approach. Upload the original gif file but also generate the mp4 as a hidden version of the file, similar to the poster image. That way, gif could still be still be filtered by the image filter but an mp4 could be present. If it worth noting, I would use the original gif as the poster image. Thoughts @swissspidy 

Thoughts @swissspidy

**Comment by swissspidy** [comment]
Thanks, I‘ll take a look next week.

We should not upload GIF files and definitely not use GIF files as poster images.

**Comment by spacedmonkey** [comment]
> We should not upload GIF files and definitely not use GIF files as poster images.

Why?

**Comment by swissspidy** [comment]
> > We should not upload GIF files and definitely not use GIF files as poster images.
> 
> Why?

The point of converting to MP4 is to improve UX and saving bandwidth by avoiding to load a 100MB GIF file. If we put the 100MB GIF file as a poster we win nothing as the browser then will losd the GIF **and** the video.

The goal of the original ticket and PR is to convert GIFs to videos and use still images as the poster and this decision still stands.

Happy to elaborate further after my vacation

**Comment by spacedmonkey** [comment]
> The goal of the original ticket and PR is to convert GIFs to videos and use still images as the poster and this decision still stands. 

I am not sure that was ever mentioned in sprint planning or in the ticket the point of this ticket was to save bandwidth. 

> avoiding to load a 100MB GIF file

In my testing, I found it hard to get a large gif. Most of the ones of sites like giphy are around 500kb - 2mb. I don't think those are unreasonable value as a poster image. It is also worth noting, the generated poster would not be animated...

**Comment by swissspidy** [comment]
I was exaggerating to get the point across.

Relevant articles:
https://web.dev/replace-gifs-with-videos/
https://paulbakaus.com/2020/01/17/gifs-must-die/

It‘s fine for the poster not to be animated.

File: assets/src/edit-story/app/media/utils/useExistingData.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+/**
+ * External dependencies
+ */
+import { useCallback } from 'react';
+/**
+ * Internal dependencies
+ */
+import useStory from '../../story/useStory';
+
+function useExistingData({ updateMedia }) {
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/media/utils/useExistingData.js]: This hook is a copy and paste from elsewhere.

File: assets/src/edit-story/app/media/utils/useProcessVideo.js [hunk 5]
```diff
@@ -32,58 +32,8 @@ function useProcessVideo({
   deleteMediaElement,
 }) {
   const { updateElementDimensions } = useUpdateElementDimensions();
-  const { updateElementsByResourceId } = useStory((state) => ({
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/media/utils/useProcessVideo.js:35]: These functions have be moved `useExistingData` hook.

**Comment by spacedmonkey** [comment]
I am going to hand this back to @swissspidy . I think it needs a little more to get this across the line. Not exactly sure what that is through.

File: web-stories.php [hunk 6]
```diff
@@ -40,8 +40,8 @@
 	exit;
 }
 
-define( 'WEBSTORIES_VERSION', '1.8.0-alpha.0' );
-define( 'WEBSTORIES_DB_VERSION', '3.0.10' );
+define( 'WEBSTORIES_VERSION', '1.9.0-alpha.0' );
+define( 'WEBSTORIES_DB_VERSION', '3.0.9' );
```
Comment (swissspidy) [inline web-stories.php]: This needs to be changed back to 3.0.10 ⚠️

File: includes/Migrations/Add_Media_Source_Gif_Conversion.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	/**
+	 * Add the gif-conversion term, to make sure it exists.
+	 *
+	 * @since 1.4.0
+	 *
+	 * @return void
+	 */
+	public function migrate() {
+		wp_insert_term( 'gif-conversion', Media\Media::STORY_MEDIA_TAXONOMY );
+		wp_insert_term( 'source-image', Media\Media::STORY_MEDIA_TAXONOMY );
```
Comment (swissspidy) [inline includes/Migrations/Add_Media_Source_Gif_Conversion.php]: Nit: technically the name of the class is now incorrect.

File: includes/Migrations/Add_Media_Source_Gif_Conversion.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+namespace Google\Web_Stories\Migrations;
+
+use Google\Web_Stories\Media;
+
+/**
+ * Class Add_Media_Source_Gif_Conversion
+ *
+ * @package Google\Web_Stories\Migrations
+ */
+class Add_Media_Source_Gif_Conversion extends Migrate_Base {
```
Comment (swissspidy) [inline includes/Migrations/Add_Media_Source_Gif_Conversion.php]: Todo: needs tests

File: assets/src/edit-story/app/media/utils/useProcessGif.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+import { useCallback } from 'react';
+import { fetchRemoteFile, isAnimatedGif } from '@web-stories-wp/media';
+
+/**
+ * Internal dependencies
+ */
+import useExistingData from './useExistingData';
+import useUpdateElementDimensions from './useUpdateElementDimensions';
+
+function useProcessGif({
```
Comment (swissspidy) [inline assets/src/edit-story/app/media/utils/useProcessGif.js]: This hook is very similar to `useProcessVideo`, we should consider merging them into one `useProcessMedia` hook

File: assets/src/edit-story/app/media/utils/useMediaUploadQueue/index.js [hunk 10]
```diff
@@ -55,8 +57,11 @@ function useMediaUploadQueue() {
     canTranscodeFile,
     transcodeVideo,
     getFirstFrameOfVideo,
+    convertGifToVideo,
   } = useFFmpeg();
 
+  const enableGifOptimization = useFeature('enableGifOptimization');
```
Comment (swissspidy) [inline assets/src/edit-story/app/media/utils/useMediaUploadQueue/index.js]: This makes it appear as if it's a function. Should rename to `isGifOptimizationEnabled` to make it clearer it's a bool flag.

File: assets/src/edit-story/app/media/utils/useExistingData.js (hunk 4 above)
Comment (swissspidy) [inline assets/src/edit-story/app/media/utils/useExistingData.js]: `useExistingData` is not really very specific. We should come up with a better name.

That said, it makes more sense to just merge the two `useProcess*` hooks into one, and removing this hook here again.

File: includes/Migrations/Add_Media_Source_Gif_Conversion.php (hunk 8 above)
Comment (spacedmonkey) [reply includes/Migrations/Add_Media_Source_Gif_Conversion.php]: Tests added.

File: includes/Migrations/Add_Media_Source_Gif_Conversion.php (hunk 7 above)
Comment (spacedmonkey) [reply includes/Migrations/Add_Media_Source_Gif_Conversion.php]: Broken into two files.

File: assets/src/edit-story/app/media/utils/useExistingData.js (hunk 4 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/media/utils/useExistingData.js]: Merged into `useProcessMedia`

File: assets/src/edit-story/app/media/utils/useMediaUploadQueue/index.js (hunk 10 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/media/utils/useMediaUploadQueue/index.js]: Changed variable name.

File: assets/src/edit-story/app/media/utils/useProcessGif.js (hunk 9 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/media/utils/useProcessGif.js]: removed

File: web-stories.php (hunk 6 above)
Comment (spacedmonkey) [reply web-stories.php]: Fixed

File: includes/Migrations/Add_Media_Source.php [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+	}
+
+	/**
+	 * Override this method.
+	 *
+	 * @since 1.9.0
+	 *
+	 * @return string
+	 */
+	public function get_term() {
```
Comment (swissspidy) [inline includes/Migrations/Add_Media_Source.php]: Since it needs to be overridden, make it abstract? Perhaps protected too.

File: assets/src/edit-story/app/media/utils/useMediaUploadQueue/index.js (hunk 10 above)
Comment (swissspidy) [reply assets/src/edit-story/app/media/utils/useMediaUploadQueue/index.js]: Apologies, I was not super clear on this one.

I meant changing the variable name, not the experiment name. Most of the feature flag names start with `enable`, so that's OK.

```js
const isGifOptimizationEnabled = useFeature('enableGifOptimization');
```

Comment (spacedmonkey) [reply assets/src/edit-story/app/media/utils/useMediaUploadQueue/index.js]: Changed it back.

File: includes/Migrations/Add_Media_Source.php (hunk 11 above)
Comment (spacedmonkey) [reply includes/Migrations/Add_Media_Source.php]: Fixed.

File: includes/Migrations/Migration_Meta_To_Term.php [hunk 12]
```diff
@@ -70,9 +70,7 @@ public function migrate() {
 	 *
 	 * @return string
 	 */
-	protected function get_post_meta_key() {
-		return '';
-	}
+	abstract protected function get_post_meta_key();
```
Comment (spacedmonkey) [inline includes/Migrations/Migration_Meta_To_Term.php:73]: Not need for this change, but thought I would this at the same time.

**Review by spacedmonkey — APPROVED** [review]

File: assets/src/edit-story/constants/media.js [hunk 13]
```diff
@@ -17,3 +17,11 @@
 const GB_IN_BYTES = 1024 * 1024 * 1024;
 
 export const MEDIA_TRANSCODING_MAX_FILE_SIZE = 2 * GB_IN_BYTES;
+
+export const MEDIA_TRANSCODED_MINE_TYPE = 'video/mp4';
```
Comment (swissspidy) [inline assets/src/edit-story/constants/media.js]:
```suggestion
export const MEDIA_TRANSCODED_MIME_TYPE = 'video/mp4';
```

File: assets/src/edit-story/components/canvas/useUploadWithPreview.js [hunk 14]
```diff
@@ -56,7 +56,11 @@ function useUploadWithPreview() {
     ({ id, resource }) => {
       updateElementDimensions({ id, resource });
 
-      if (resource.type === 'video' && !resource.local && !resource.posterId) {
+      if (
+        ['video', 'gif'].includes(resource.type) &&
```
Comment (swissspidy) [inline assets/src/edit-story/components/canvas/useUploadWithPreview.js:60]: Note to self: this also affects Tenor GIFs.

File: assets/src/edit-story/app/media/utils/getResourceFromLocalFile.js (hunk 2 above)
Comment (swissspidy) [reply assets/src/edit-story/app/media/utils/getResourceFromLocalFile.js]: Because it's not needed

**Comment by swissspidy** [comment]
@spacedmonkey I added some data migration here. If you could review again that would be great.

In testing, make sure both new and old Tenor GIFs still work as expected.

**Comment by spacedmonkey** [comment]
There is a performance regression between main and this branch. See an example video dragged the same video into canvas. 

 #### Main

https://user-images.githubusercontent.com/237508/124271165-0810cb00-db35-11eb-9f31-d70dd028677b.mov

#### This branch

https://user-images.githubusercontent.com/237508/124271281-2f679800-db35-11eb-9465-30d28857ab9d.mov

At a guess, I believe the normal poster generation is failing and it is failing back to poster generation via ffmpeg. I believe it is related to this change. https://github.com/google/web-stories-wp/pull/7528/commits/471bb13a45f690c009a8dda5a1985b1d11eac8aa .

File: assets/src/edit-story/constants/media.js [hunk 15]
```diff
@@ -17,5 +17,14 @@
 const GB_IN_BYTES = 1024 * 1024 * 1024;
 
 export const MEDIA_TRANSCODING_MAX_FILE_SIZE = 2 * GB_IN_BYTES;
+
 export const MEDIA_POSTER_IMAGE_EXT = 'jpeg';
 export const MEDIA_POSTER_IMAGE_MIME_TYPE = 'image/jpeg';
+
+export const MEDIA_TRANSCODED_MIME_TYPE = 'video/mp4';
+export const MEDIA_TRANSCODED_FILE_TYPE = 'mp4';
```
Comment (spacedmonkey) [inline assets/src/edit-story/constants/media.js:25]: This should match 
```
export const MEDIA_POSTER_IMAGE_EXT = 'jpeg';
```

Either one should be change to match the other.

Comment (swissspidy) [reply assets/src/edit-story/constants/media.js:25]: You mean renaming to `MEDIA_TRANSCODED_VIDEO_MIME_TYPE`/`MEDIA_TRANSCODED_VIDEO_FILE_TYPE`? Or what exactly should match?

**Comment by swissspidy** [comment]
> At a guess, I believe the normal poster generation is failing and it is failing back to poster generation via ffmpeg. I believe it is related to this change. [471bb13](https://github.com/google/web-stories-wp/commit/471bb13a45f690c009a8dda5a1985b1d11eac8aa) .

Could you try reverting it locally to see if it makes a difference? Not too attached to it , so if it doesn't work I can undo it. (although I'd love to tidy this up one day)

File: assets/src/edit-story/constants/media.js (hunk 15 above)
Comment (spacedmonkey) [reply assets/src/edit-story/constants/media.js:25]: The naming of `MEDIA_POSTER_IMAGE_EXT` / `MEDIA_TRANSCODED_FILE_TYPE` do not match. Where as `MEDIA_POSTER_IMAGE_MIME_TYPE` and `MEDIA_TRANSCODED_MIME_TYPE` does.

**Comment by spacedmonkey** [comment]
> Could you try reverting it locally to see if it makes a difference? Not too attached to it , so if it doesn't work I can undo it. (although I'd love to tidy this up one day)

I reverting back to 75af1096a7650c4d87cb275e2eb0db4317f230bb and I am seeing the behaviour found in main.  `createFileReader` is needed. 


https://user-images.githubusercontent.com/237508/124455119-1bb07180-dd81-11eb-9634-3e2be8e2aee2.mov

**Review by spacedmonkey — DISMISSED** [review]
Last commit needs to reverted.

**Comment by spacedmonkey** [comment]
@swissspidy Conflicts need to be resolved here.

File: assets/src/edit-story/constants/media.js (hunk 15 above)
Comment (swissspidy) [reply assets/src/edit-story/constants/media.js:25]: Renaming `MEDIA_POSTER_IMAGE_EXT` to `MEDIA_POSTER_IMAGE_FILE_TYPE`

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
Niced a weird bug in my testing. See video. Missing poster image. 

https://user-images.githubusercontent.com/237508/124602102-6c4acc00-de60-11eb-89a7-9cc89c567b28.mov

**Comment by swissspidy** [comment]
> Niced a weird bug in my testing. See video. Missing poster image.


Good catch! fe1a26ec34d5df15d3bec49259b5c327b9b677c4 should fix that.

**Comment by spacedmonkey** [comment]
This code converted a 97MB Gif to a 2mb mp4. Win!

**Review by spacedmonkey — APPROVED** [review]
👍
